drivers/sensor/bh1750/bh1750.c

```diff
@@ -0,0 +1,231 @@
+#warning "BH1750 driver enabled without any devices"
+#endif
+
+#define BH1750_POWER_DOWN                          0x00
+#define BH1750_POWER_ON                            0x01
+#define BH1750_RESET                               0x07
+#define BH1750_CONTINUOUSLY_HIGH_RESOLUTION_MODE   0x10
+#define BH1750_CONTINUOUSLY_HIGH_RESOLUTION_MODE_2 0x11
+#define BH1750_CONTINUOUSLY_LOW_RESOLUTION_MODE    0x13
+#define BH1750_ONE_TIME_HIGH_RESOLUTION_MODE       0x20
+#define BH1750_ONE_TIME_HIGH_RESOLUTION_MODE_2     0x21
+#define BH1750_ONE_TIME_LOW_RESOLUTION_MODE        0x23
+#define BH1750_CHANGE_MEASUREMENT_TIME_HIGH        0x40
+#define BH1750_CHANGE_MEASUREMENT_TIME_LOW         0x60
+
+struct bh1750_data {
+	uint16_t raw_val;
+};
+
+struct bh1750_config {
+	struct i2c_dt_spec i2c;
+};
+
+static inline int bh1750_is_ready(const struct device *dev)
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return device_is_ready(cfg->i2c.bus) ? 0 : -ENODEV;
+}
+
+static inline int bh1750_read(const struct device *dev, uint8_t *buf, int size)
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_read_dt(&cfg->i2c, buf, size);
+}
+
+static inline int bh1750_write(const struct device *dev, uint8_t val)
+{
+	const struct bh1750_config *cfg = dev->config;
+	uint8_t buf = val;
+
+	return i2c_write_dt(&cfg->i2c, &buf, sizeof(buf));
+}
+
+static int bh1750_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct bh1750_data *data = dev->data;
+	uint8_t buf[2];
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+	(void)pm_device_state_get(dev, &state);
+	/* Do not allow sample fetching from suspended state */
+	if (state == PM_DEVICE_STATE_SUSPENDED) {
+		return -EIO;
+	}
+#endif
+
+	ret = bh1750_write(dev, BH1750_ONE_TIME_HIGH_RESOLUTION_MODE);
+	if (ret < 0) {
+		LOG_DBG("One time high resolution mode failed: %d", ret);
+		return ret;
+	}
+
+	/* Wait for the measure to be ready */
+	k_sleep(K_MSEC(180)); /* Max for H-Resolution Mode */
+
+	ret = bh1750_read(dev, buf, sizeof(buf));
+	if (ret < 0) {
+		LOG_DBG("Read failed: %d", ret);
+		return ret;
+	}
+
+	data->raw_val = sys_get_be16(&buf[0]);
+
+	return 0;
+}
+
+static int bh1750_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bh1750_data *data = dev->data;
+	double tmp;
+	int ret;
+
+	switch (chan) {
+	case SENSOR_CHAN_LIGHT:
+		/*
+		 * The documentation says the following about the calculation
+		 * at section Measurement sequence example from "Write
+		 * instruction" to "Read measurement result" p.17:
+		 *
+		 * How to calculate when the data High Byte is "10000011" and
+		 * Low Byte is "10010000"
+		 *
+		 * (2^15 + 2^9 + 2^8 + 2^7 + 2^4) / 1.2 = 28067 [lx]
+		 *
+		 * Which means:
+		 *
+		 * lx = Slx / 1.2
+		 */
+		tmp = data->raw_val / 1.2;
+		ret = sensor_value_from_double(val, tmp);
```

Comment:
@MaureenHelm: as for [51269](https://github.com/zephyrproject-rtos/zephyr/pull/51269#discussion_r997168142)

> You don't need floating point math for this

do you prefer the former solution with long long unsigned?

``` c
static int bh1750_channel_get(const struct device *dev,
			      enum sensor_channel chan,
			      struct sensor_value *val)
{
	struct bh1750_data *data = dev->data;
	long long unsigned tmp;

	switch (chan) {
	case SENSOR_CHAN_LIGHT:
		tmp = 1000000UL * data->raw_val / 1.2;
		val->val1 = tmp / 1000000;
		val->val2 = tmp % 1000000;
		break;
	default:
		return -EINVAL;
	}

	return 0;
}
```

Reply:
@MaureenHelm do you want me to switch back to `long long unsigned` for that driver as well?